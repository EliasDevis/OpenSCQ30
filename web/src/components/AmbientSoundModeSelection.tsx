import { Box, Typography } from "@mui/material";
import { AmbientSoundMode } from "../../wasm/pkg/openscq30_web_wasm";
import { ToggleButtonRow } from "./ToggleButtonRow";

type Props = {
  value: AmbientSoundMode;
  onValueChanged: (newValue: AmbientSoundMode) => void;
};

export function AmbientSoundModeSelection(props: Props) {
  return (
    <Box>
      <Typography>Ambient Sound Mode</Typography>
      <ToggleButtonRow
        value={props.value}
        onValueChanged={props.onValueChanged}
        values={[
          {
            value: AmbientSoundMode.NoiseCanceling,
            displayText: "Noise Canceling",
          },
          {
            value: AmbientSoundMode.Transparency,
            displayText: "Transparency",
          },
          {
            value: AmbientSoundMode.Normal,
            displayText: "Normal",
          },
        ]}
      />
    </Box>
  );
}
