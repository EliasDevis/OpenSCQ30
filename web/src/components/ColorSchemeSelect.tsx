import { DarkMode, LightMode } from "@mui/icons-material";
import { IconButton, Menu, MenuItem, useColorScheme } from "@mui/material";
import { useState } from "react";
import { useTranslation } from "react-i18next";

export function ColorSchemeSelect() {
  const { t } = useTranslation();
  const { mode, setMode, systemMode } = useColorScheme();
  const currentScheme = mode == "system" ? systemMode : mode;

  const [anchorEl, setAnchorEl] = useState<HTMLElement | null>(null);
  const isOpen = !!anchorEl;

  function close() {
    setAnchorEl(null);
  }

  const options = [
    { key: "system", translation: t("colorScheme.system") },
    { key: "light", translation: t("colorScheme.light") },
    { key: "dark", translation: t("colorScheme.dark") },
  ] as const;

  function selectItem(key: (typeof options)[number]["key"]) {
    setMode(key);
    close();
  }

  return (
    <div>
      <IconButton
        id="color-scheme-button"
        aria-label={t("colorScheme.colorScheme")}
        onClick={(event) => setAnchorEl(event.currentTarget)}
      >
        {currentScheme == "dark" ? <LightMode /> : <DarkMode />}
      </IconButton>
      <Menu
        anchorEl={anchorEl}
        open={isOpen}
        onClose={close}
        MenuListProps={{
          "aria-labelledby": "color-scheme-button",
        }}
      >
        {options.map((option, i) => (
          <MenuItem
            key={i}
            onClick={() => selectItem(option.key)}
            selected={mode == option.key}
          >
            {option.translation}
          </MenuItem>
        ))}
      </Menu>
    </div>
  );
}
