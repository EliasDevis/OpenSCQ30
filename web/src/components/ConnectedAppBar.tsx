import { AppBar, Button, Toolbar, Typography } from "@mui/material";
import { useTranslation } from "react-i18next";

type Props = {
  deviceName: string;
  onDisconnectClick: () => void;
};

export function ConnectedAppBar(props: Props) {
  const { t } = useTranslation();
  return (
    <AppBar position="absolute">
      <Toolbar>
        <Typography
          component="h1"
          variant="h6"
          color="inherit"
          noWrap
          sx={{ flexGrow: 1 }}
        >
          {props.deviceName}
        </Typography>
        <Button color="inherit" onClick={() => props.onDisconnectClick()}>
          {t("device.disconnect")}
        </Button>
      </Toolbar>
    </AppBar>
  );
}
