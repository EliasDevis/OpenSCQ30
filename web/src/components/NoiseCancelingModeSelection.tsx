import { Box, Typography } from "@mui/material";
import { NoiseCancelingMode } from "../../wasm/pkg/openscq30_web_wasm";
import { ToggleButtonRow } from "./ToggleButtonRow";

type Props = {
  value: NoiseCancelingMode;
  onValueChanged: (newValue: NoiseCancelingMode) => void;
};

export function NoiseCancelingModeSelection(props: Props) {
  return (
    <Box>
      <Typography>Ambient Sound Mode</Typography>
      <ToggleButtonRow
        value={props.value}
        onValueChanged={props.onValueChanged}
        values={[
          {
            value: NoiseCancelingMode.Transport,
            displayText: "Transport",
          },
          {
            value: NoiseCancelingMode.Outdoor,
            displayText: "Outdoor",
          },
          {
            value: NoiseCancelingMode.Indoor,
            displayText: "Indoor",
          },
        ]}
      />
    </Box>
  );
}
