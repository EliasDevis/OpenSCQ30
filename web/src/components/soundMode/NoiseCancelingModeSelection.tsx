import { Box, Typography } from "@mui/material";
import { NoiseCancelingMode } from "../../../wasm/pkg/openscq30_web_wasm";
import { ToggleButtonRow } from "./ToggleButtonRow";
import { useTranslation } from "react-i18next";
import React from "react";

interface Props {
  value: NoiseCancelingMode;
  onValueChanged: (newValue: NoiseCancelingMode) => void;
}

export const NoiseCancelingModeSelection = React.memo(function (props: Props) {
  const { t } = useTranslation();
  return (
    <Box>
      <Typography>{t("noiseCancelingMode.noiseCancelingMode")}</Typography>
      <ToggleButtonRow
        value={props.value}
        onValueChanged={props.onValueChanged}
        values={[
          {
            value: NoiseCancelingMode.Transport,
            displayText: t("noiseCancelingMode.transport"),
          },
          {
            value: NoiseCancelingMode.Outdoor,
            displayText: t("noiseCancelingMode.outdoor"),
          },
          {
            value: NoiseCancelingMode.Indoor,
            displayText: t("noiseCancelingMode.indoor"),
          },
        ]}
      />
    </Box>
  );
});
