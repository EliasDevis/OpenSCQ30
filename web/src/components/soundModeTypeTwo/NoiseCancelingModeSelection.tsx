import { Box, Typography } from "@mui/material";
import { ToggleButtonRow, ToggleButtonValues } from "./ToggleButtonRow";
import { useTranslation } from "react-i18next";
import React from "react";
import { SoundModesTypeTwo } from "../../libTypes/DeviceState";
import { useFilterNulls } from "../../hooks/useFilterNulls";

interface Props {
  value: SoundModesTypeTwo["noiseCancelingMode"];
  onValueChanged: (newValue: SoundModesTypeTwo["noiseCancelingMode"]) => void;
}

export const NoiseCancelingModeSelection = React.memo(function ({
  value,
  onValueChanged,
}: Props) {
  const { t } = useTranslation();
  // Don't allow deselecting the button
  const onValueChangedNotNull = useFilterNulls(onValueChanged);

  const values: ToggleButtonValues<SoundModesTypeTwo["noiseCancelingMode"]> = [
    {
      value: "adaptive",
      label: t("noiseCancelingMode.adaptive"),
    },
    {
      value: "manual",
      label: t("noiseCancelingMode.manual"),
    },
  ];

  return (
    <Box>
      <Typography>{t("soundModes.noiseCancelingMode")}</Typography>
      <ToggleButtonRow
        value={value}
        onValueChanged={onValueChangedNotNull}
        values={values}
      />
    </Box>
  );
});
