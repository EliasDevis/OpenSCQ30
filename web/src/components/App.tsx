import {
  Box,
  Container,
  ThemeProvider,
  Toolbar,
  createTheme,
  useMediaQuery,
} from "@mui/material";
import { useMemo, useState } from "react";
import { SoundcoreDevice, selectDevice } from "../bluetooth/SoundcoreDevice";
import { DeviceSettings } from "./DeviceSettings";
import { ConnectedAppBar } from "./ConnectedAppBar";
import { DisconnectedAppBar } from "./DisconnectedAppBar";
import { HomePage } from "./HomePage";

function App() {
  const [device, setDevice] = useState<SoundcoreDevice>();
  const prefersDarkMode = useMediaQuery("(prefers-color-scheme: dark)");
  const theme = useMemo(
    () =>
      createTheme({
        palette: {
          mode: prefersDarkMode ? "dark" : "light",
        },
      }),
    [prefersDarkMode]
  );

  async function connect() {
    const device = await selectDevice();
    setDevice(device);
  }
  function disconnect() {
    device?.disconnect();
    setDevice(undefined);
  }

  return (
    <ThemeProvider theme={theme}>
      <Box
        sx={{
          display: "flex",
          backgroundColor: theme.palette.background.default,
          minHeight: "100vh",
        }}
        color="text.primary"
      >
        {device ? (
          <ConnectedAppBar
            deviceName={device.name ?? "Unknown device"}
            onDisconnectClick={() => disconnect()}
          />
        ) : (
          <DisconnectedAppBar onSelectDeviceClick={() => connect()} />
        )}
        <Box component="main" sx={{ flexGrow: 1 }}>
          <Toolbar />
          <Container maxWidth="sm" sx={{ my: 2 }}>
            {device ? <DeviceSettings device={device} /> : <HomePage />}
          </Container>
        </Box>
      </Box>
    </ThemeProvider>
  );
}

export default App;
