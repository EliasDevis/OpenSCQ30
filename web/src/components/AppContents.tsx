import { Box, Container, Toolbar, useTheme } from "@mui/material";
import { useCallback, useState } from "react";
import { useUpdateAvailableToast } from "../hooks/useUpdateAvailableToast";
import { ConnectedAppBar } from "./ConnectedAppBar";
import { DisconnectedAppBar } from "./DisconnectedAppBar";
import { HomePage } from "./HomePage";
import { DeviceSettings } from "./deviceSettings/DeviceSettings";
import { Device, selectDemoDevice, selectDevice } from "../bluetooth/Device";
import { LoadingScreen } from "./LoadingScreen";

export function AppContents() {
  const [device, setDevice] = useState<Device>();
  const [isLoading, setLoading] = useState(false);
  const isDemoMode = localStorage.getItem("openscq30:demoMode") == "true";
  const theme = useTheme();
  useUpdateAvailableToast();

  const connect = useCallback(
    (isFiltered: boolean) => {
      setLoading(true);
      (isDemoMode ? selectDemoDevice : selectDevice)(isFiltered)
        .then(setDevice)
        .catch((err) => {
          setLoading(false);
          // Ignore error if the user canceled the device selection popup
          if (!(err instanceof DOMException) || err.name != "NotFoundError") {
            console.error(err);
          }
        });
    },
    [isDemoMode],
  );

  const connectFiltered = useCallback(() => {
    connect(true);
  }, [connect]);
  const connectUnfiltered = useCallback(() => {
    connect(false);
  }, [connect]);

  const disconnect = useCallback(() => {
    device?.destroy();
    setDevice(undefined);
    setLoading(false);
  }, [device]);

  return (
    <>
      <Box
        sx={{
          display: "flex",
          backgroundColor: theme.palette.background.default,
          minHeight: "100vh",
        }}
        color="text.primary"
      >
        {device ? (
          <ConnectedAppBar
            deviceName={device.name ?? "Unknown device"}
            onDisconnectClick={disconnect}
          />
        ) : (
          <DisconnectedAppBar
            onSelectDeviceClick={connectFiltered}
            onSelectDeviceUnfilteredClick={connectUnfiltered}
            showSelectDeviceButton={!!navigator.bluetooth || isDemoMode}
          />
        )}
        <Box component="main" sx={{ flexGrow: 1 }}>
          <Toolbar />
          <Container sx={{ my: 2 }}>
            {device ? (
              <DeviceSettings device={device} disconnect={disconnect} />
            ) : isLoading ? (
              <LoadingScreen />
            ) : (
              <HomePage />
            )}
          </Container>
        </Box>
      </Box>
    </>
  );
}

export default AppContents;
