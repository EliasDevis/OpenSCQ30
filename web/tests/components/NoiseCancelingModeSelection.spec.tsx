import { fireEvent, render } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";
import { NoiseCancelingModeSelection } from "../../src/components/NoiseCancelingModeSelection";
import { NoiseCancelingMode } from "../../wasm/pkg/openscq30_web_wasm";

describe("NoiseCancelingModeSelection", () => {
  it("should render", () => {
    const renderResult = render(
      <NoiseCancelingModeSelection
        value={NoiseCancelingMode.Indoor}
        onValueChanged={() => {
          throw new Error("Function not implemented.");
        }}
      />
    );

    expect(renderResult.asFragment()).toMatchSnapshot();
  });

  it("should fire onValueChanged when clicked", () => {
    const setMode = vi.fn();
    const renderResult = render(
      <NoiseCancelingModeSelection
        value={NoiseCancelingMode.Indoor}
        onValueChanged={setMode}
      />
    );

    fireEvent.click(renderResult.getByText("Outdoor"));

    expect(setMode).toHaveBeenCalledWith(NoiseCancelingMode.Outdoor);
  });
});
