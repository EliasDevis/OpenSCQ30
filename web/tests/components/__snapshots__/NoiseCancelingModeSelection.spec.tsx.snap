// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NoiseCancelingModeSelection > should render 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-0"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
    >
      noiseCancelingMode.noiseCancelingMode
    </p>
    <div
      class="MuiToggleButtonGroup-root css-1yt1mvy-MuiToggleButtonGroup-root"
      role="group"
    >
      <button
        aria-pressed="false"
        class="MuiButtonBase-root MuiToggleButton-root MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal css-1psryzy-MuiButtonBase-root-MuiToggleButton-root"
        tabindex="0"
        type="button"
        value="0"
      >
        noiseCancelingMode.transport
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        aria-pressed="false"
        class="MuiButtonBase-root MuiToggleButton-root MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal css-1psryzy-MuiButtonBase-root-MuiToggleButton-root"
        tabindex="0"
        type="button"
        value="1"
      >
        noiseCancelingMode.outdoor
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        aria-pressed="true"
        class="MuiButtonBase-root MuiToggleButton-root Mui-selected MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal css-1psryzy-MuiButtonBase-root-MuiToggleButton-root"
        tabindex="0"
        type="button"
        value="2"
      >
        noiseCancelingMode.indoor
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
